src/main/java/org/opentestsystem/ap/item/history/ItemHistoryEventHandler.java

```
@@ -0,0 +1,182 @@
+     *
+     * @param item       The item to generate SAAIF for.
+     * @param itemFolder The folder to generate the data in.
+     */
+    private void generateSaaifContent(Item item, Path itemFolder) {
+        ItemContext itemContext = new ItemContext(
+            item.getId(),
+            item.getLegacyId(this.itemBankProperties.getBankKey()),
+            this.itemBankProperties.getBankKey(),
+            this.appAssembler,
+            itemFolder
+        );
+        ModelTransformer transformer = transformerFactory.newTransformer(item.getType(), this.imageGenerator);
+        transformer.generateSaaifFiles(item, itemBankProperties.getBankKey(), itemContext);
+    }
+
+    /**
+     * Downloads the attachments associated with the item.  The attachments version is used if provided.  It is used to
+     * download a specific version of the attachment.  If the version is not provided the latest version of the
+     * attachments are downloaded individually.  This is a fallback behavior as it is expected the version ID is
+     * provided as it should have been generate prior to the item event being sent.
+     *
+     * @param itemId              The item to download the attachments for.
+     * @param attachmentVersionId The version of the attachments to download.
+     * @return The path were the attachments have been downloaded.
+     */
+    private Path downloadAttachments(String itemId, String attachmentVersionId) {
+        // use 'localBaseDir' so the clean up process takes care of deleting these files
+        Path downloadFolder = Paths.get(itemBankProperties.getLocalBaseDir(), idGenerator.generateTransactionId());
+        if (StringUtils.isNotBlank(attachmentVersionId)) {
+            dataStoreAttachmentManager.downloadVersionedAttachments(itemId, attachmentVersionId, downloadFolder);
+        } else {
+            dataStoreAttachmentManager.downloadAttachments(itemId, BRANCH_MASTER, downloadFolder);
```

Should there a backup or should this outright fail?  My concern is that if the attachment version id is required we shouldn't silently use master?

---

Let's fail.  When committing changes on an item attachment version ID is required.  If it isn't present something is wrong.  You are right silently using master is a bad idea.

Let's discuss this tomorrow at parking lot.  I want to describe how an item **commit** event produces a new attachment version but an item **patch** event does not, it just reuses what is already set on the item if anything at all.  Maybe commit and patch are handled differently by the history service where I have those two events handled identically.